### PR TITLE
Convert MarkdownColor to class syntax

### PR DESCRIPTION
**Pull request by locks**

### CODE REVIEW

File: lib/utilities/markdown-color.js [hunk 1]
```diff
@@ -33,109 +31,117 @@ module.exports = MarkdownColor;
...
-  let colorTokens = unshiftValue(colors.concat(backgroundColors).map(getToken), {}).reduce(setToken);
-  return merge(colorTokens, defaultTokens);
-};
+  /*
+    Generate default style tokens
+
+    @method generateTokens
+    @return {Object}
+  */
+  generateTokens(renderer) {
```
Comment (twokul) [inline lib/utilities/markdown-color.js]: it looks like this method is internal to the class; at the very least we should add `@private` to the docs and change the name to `_ generateTokens` imo. I'd probably go as far as extracting this as a util function and not putting it on the class.

File: lib/utilities/markdown-color.js [hunk 2]
```diff
@@ -33,109 +31,117 @@ module.exports = MarkdownColor;
...
+  }
 
+  /*
+    Looks up multiple styles to apply to the rendered output
+    @method renderStylesFactory
+    @param {Object} [renderer] should match chalk.js format
+    @param {String, Array} [styleNames]
+    @return {Function} Function that will wrap the input string with styling
+  */
+  renderStylesFactory(renderer, styleNames) {
```
Comment (twokul) [inline lib/utilities/markdown-color.js]: same here

File: lib/utilities/markdown-color.js [hunk 3]
```diff
@@ -33,109 +31,117 @@ module.exports = MarkdownColor;
...
-  };
+  /*
+    Parse markdown and output as string
+
+    @method render
+    @param {String} [string]
+    @param {Object} [options]
+    @return {String}
+  */
+  render(string, options) {
```
Comment (twokul) [inline lib/utilities/markdown-color.js]: it doesn't seem like `render` is used anywhere?

Comment (locks) [inline lib/utilities/markdown-color.js]: it's being specifically tested for

File: lib/utilities/markdown-color.js [hunk 4]
```diff
@@ -33,109 +31,117 @@ module.exports = MarkdownColor;
...
+    this.markdown = markdown().use(terminal, markdownOptions);
+    this.tokens = merge(tokens, optionTokens);
   }
 
-  return this.render(file, options);
-};
+  /*
+    Lookup markdown file and render contents
+
+    @private
```
Comment (Turbo87) [inline lib/utilities/markdown-color.js]: why is this private now? seems like an unrelated change

File: lib/utilities/markdown-color.js [hunk 5]
```diff
@@ -173,26 +180,30 @@ function getColorTokenRegex(name, options) {
...
 
 function escapeToken(match, pattern) {
   let output = pattern.replace(/\~/g, '');
+
   return `<${output}>`;
 }
 
 function unshiftValue(array, value) {
   array.unshift(value);
+
```
Comment (Turbo87) [inline lib/utilities/markdown-color.js]: why the extra whitespace?

File: lib/utilities/markdown-color.js (hunk 1 above)
Comment (Turbo87) [inline lib/utilities/markdown-color.js]: @twokul the point of the PR is converting to class syntax. IMHO it shouldn't change any visibility levels, behaviors, or rename anything.

File: lib/utilities/markdown-color.js (hunk 5 above)
Comment (locks) [inline lib/utilities/markdown-color.js]: I find it tidier when returns are separated.

File: lib/utilities/markdown-color.js [hunk 6]
```diff
@@ -177,7 +177,7 @@ function getToken(name, options) {
     name,
     token: `<${name}></${name}>`,
     pattern: getColorTokenRegex(name),
-    render: renderStylesFactory(renderer, name),
+    render: this.renderStylesFactory(renderer, name),
```
Comment (Turbo87) [inline lib/utilities/markdown-color.js]: how does this function get access to `this`?

Comment (locks) [inline lib/utilities/markdown-color.js]: Right, this was a tricky one. I'm binding `this`: https://github.com/ember-cli/ember-cli/pull/7259/files#diff-f6db1f1718f3ab9018f41d583b46f03dR111. Should I pull it into the class? I didn't want to change the surface API.

**Comment by locks** [comment]
🏓
